src/box/txn_event_trigger.c

```diff
@@ -0,0 +1,340 @@
+		txn_event->space_event = space_event;
+		break;
+	case TXN_EVENT_ONE_SPACE:
+		if (txn_event->space_id != space_id(space)) {
+			txn_event->mode = TXN_EVENT_MULTIPLE_SPACES;
+			txn_event->space_id = 0;
+			txn_event->space_event = NULL;
+		}
+		break;
+	case TXN_EVENT_MULTIPLE_SPACES:
+		break;
+	default:
+		unreachable();
+	}
+}
+
+enum { SPACE_ID_FILTER_ALL_SPACES = -1 };
+
+/**
+ * State of the iterator, which is passed to txn_iterator_next().
+ */
+struct txn_iterator_state {
+	/** Request number, starting from 1. */
+	uint32_t req_num;
+	/** Current statement of the transaction. */
+	struct txn_stmt *stmt;
+	/**
+	 * Iterate only over statements with the given space id.
+	 * Or set to SPACE_ID_FILTER_ALL_SPACES for all spaces.
+	 */
+	int64_t space_id_filter;
+};
+
+/**
+ * The iterator goes through every statement of the transaction.
+ */
+static int
+txn_iterator_next(struct func_adapter *func, struct func_adapter_ctx *ctx,
+		  void *state)
+{
+	struct txn_iterator_state *txn_state =
+		(struct txn_iterator_state *)state;
+	struct txn_stmt *stmt = txn_state->stmt;
+
+	/* Skip unwanted spaces if space id filter is set. */
+	if (txn_state->space_id_filter != SPACE_ID_FILTER_ALL_SPACES) {
+		while (stmt != NULL &&
+		       space_id(stmt->space) != txn_state->space_id_filter) {
+			stmt = stailq_next_entry(stmt, next);
+		}
+	}
+
+	if (stmt == NULL)
+		return 0;
+	/*
+	 * The iterator returns 4 values:
+	 *  1. An ordinal request number;
+	 *  2. The old value of the tuple;
+	 *  3. The new value of the tuple;
+	 *  4. The ID of the space.
+	 */
+	func_adapter_push_double(func, ctx, txn_state->req_num++);
+	if (stmt->old_tuple != NULL)
+		func_adapter_push_tuple(func, ctx, stmt->old_tuple);
+	else
+		func_adapter_push_null(func, ctx);
+	if (stmt->new_tuple != NULL)
+		func_adapter_push_tuple(func, ctx, stmt->new_tuple);
+	else
+		func_adapter_push_null(func, ctx);
+	func_adapter_push_double(func, ctx, space_id(stmt->space));
+
+	txn_state->stmt = stailq_next_entry(stmt, next);
+	return 0;
+}
+
+/**
+ * Run triggers registered in `txn' for the `event'.
+ * `stmt' points to the first statement of a transaction or a savepoint.
+ * `can_abort' is set to true if the trigger can abort the transaction.
+ * `space_id' is passed to the iterator to filter spaces by the ID.
+ */
+static int
+run_triggers_general(struct txn *txn, struct txn_stmt *stmt,
+		     struct event *event, bool can_abort, int64_t space_id)
+{
+	int rc = 0;
+	const char *name = NULL;
+	struct func_adapter *trigger = NULL;
+	struct func_adapter_ctx ctx;
+	struct event_trigger_iterator it;
+	event_trigger_iterator_create(&it, event);
+
+	while (rc == 0 && event_trigger_iterator_next(&it, &trigger, &name)) {
+		if (can_abort) {
+			/*
+			 * The transaction could be aborted while the previous
+			 * trigger was running (e.g. if the trigger-function
+			 * yielded or failed).
+			 */
+			rc = txn_check_can_continue(txn);
+			if (rc != 0)
+				break;
+		}
```

Comment:
I think that we should always call `txn_check_can_continue`, but in the case of rollback triggers we should call `panic` or just return an error - it will be logged.

Reply:
Makes sense.